src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -225,20 +222,20 @@ private async Task RecomputeTagsAsync(bool highPriority, CancellationToken cance
                     var oldState = this.State;
 
                     var textChangeRange = this.AccumulatedTextChanges;
-                    this.AccumulatedTextChanges = null;
+                    var subjectBufferVersion = _subjectBuffer.CurrentSnapshot.Version.VersionNumber;
 
-                    // Technically not necessary since we ConfigureAwait(false) right above this.  But we want to ensure
-                    // we're always moving to the threadpool here in case the above code ever changes.
                     await TaskScheduler.Default;
 
-                    cancellationToken.ThrowIfCancellationRequested();
+                    if (cancellationToken.IsCancellationRequested)
+                        return;
 
                     // Create a context to store pass the information along and collect the results.
                     var context = new TaggerContext<TTag>(
                         oldState, spansToTag, caretPosition, textChangeRange, oldTagTrees);
                     await ProduceTagsAsync(context, cancellationToken).ConfigureAwait(false);
 
-                    cancellationToken.ThrowIfCancellationRequested();
+                    if (cancellationToken.IsCancellationRequested)
```

Comment:
>if (cancellationToken.IsCancellationRequested) [](http://example.com/codeflow?start=20&length=46)

Would it make sense to do this right before the SwitchToMainThreadAsync(cancellationToken) a couple lines down to reduce the probably of the exception being thrown?

Reply:
this ct check seems highly unlikely to be relevant.  it feels like it would only be if in the actual work to compute tags we were *not* checking CT, and we only just slipped in something here.

Reply:
I don't think this placement is a significant contributor to the problem, but it made sense to update for consistency with the others in the same method.

Reply:
I assume the most relevant points are the 3 thread switches and the DelayWhileNonVisibleAsync call. Of those 4, only the background thread switch doesn't throw if the ct is cancelled, right?

Reply:
For my understanding, there were 45K exceptions in this method and the calling method had 110K, right? From the old code, I'm guessing the caller's extra 65K were all due to the main thread switch, which was moved into this method, which means that it would have 110K exceptions now, and the caller would have 0 since it's no longer building the async state machine. Definitely a win, I was just curious whether my understanding is right and whether we can do anything relatively simple to improve it.

I assume calling SwitchToMainThread without using the cancellationToken directly and checking it after or using NoThrowAwaitableInternal down in DelayWhileNonVisibleAsync wouldn't help.

Reply:
Surprisingly, an even dumber question: If you move the SwitchToMainThreadAsync call into RecomputeTagsAsync, why not just put a big old try/catch for TaskCancelledException and be done with it?

Reply:
➡️ This pull request does not aim to address the exceptions thrown by SwitchToMainThreadAsync in these methods.

> why not just put a big old try/catch for (Operation/Task)CanceledException and be done with it

Oddly this would undermine the compiler work to avoid throwing in these cases (exceptions will still be thrown if you `await` inside a `try` block that contains one or more `catch` clauses).

Reply:
➡️ After 17.7 Preview 2 (or 3) ships, we can use the changes from https://github.com/microsoft/vs-threading/pull/1194 to eliminate the remaining exceptions here